polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -152,105 +153,109 @@ async fn run<Context>(
-						let validation_host = validation_host.clone();
-
-						async move {
-							let _timer = metrics.time_validate_from_chain_state();
-							let res = validate_from_chain_state(
-								&mut sender,
-								validation_host,
-								candidate_receipt,
-								pov,
-								executor_params,
-								timeout,
-								&metrics,
-							)
-							.await;
-
-							metrics.on_validation_event(&res);
-							let _ = response_sender.send(res);
-						}
-					};
-
-					ctx.spawn("validate-from-chain-state", bg.boxed())?;
-				},
-				CandidateValidationMessage::ValidateFromExhaustive(
-					persisted_validation_data,
-					validation_code,
-					candidate_receipt,
-					pov,
-					executor_params,
-					timeout,
-					response_sender,
-				) => {
-					let bg = {
-						let metrics = metrics.clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let _timer = metrics.time_validate_from_exhaustive();
-							let res = validate_candidate_exhaustive(
-								validation_host,
-								persisted_validation_data,
-								validation_code,
-								candidate_receipt,
-								pov,
-								executor_params,
-								timeout,
-								&metrics,
-							)
-							.await;
-
-							metrics.on_validation_event(&res);
-							let _ = response_sender.send(res);
-						}
-					};
-
-					ctx.spawn("validate-from-exhaustive", bg.boxed())?;
-				},
-				CandidateValidationMessage::PreCheck(
-					relay_parent,
-					validation_code_hash,
-					response_sender,
-				) => {
-					let bg = {
-						let mut sender = ctx.sender().clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let precheck_result = precheck_pvf(
-								&mut sender,
-								validation_host,
-								relay_parent,
-								validation_code_hash,
-							)
-							.await;
-
-							let _ = response_sender.send(precheck_result);
-						}
-					};
-
-					ctx.spawn("candidate-validation-pre-check", bg.boxed())?;
-				},
+	let mut res = Ok(());
+	let sender = ctx.sender().to_owned();
+
+	let read_stream = stream::poll_fn(|c| loop {
+		match ctx.recv().poll_unpin(c) {
+			Poll::Ready(Ok(FromOrchestra::Signal(OverseerSignal::Conclude))) =>
+				return Poll::Ready(None),
+			Poll::Pending => return Poll::Pending,
+			Poll::Ready(Ok(FromOrchestra::Signal(_))) => continue,
+			Poll::Ready(Ok(FromOrchestra::Communication { msg })) => return Poll::Ready(Some(msg)),
+			Poll::Ready(Err(e)) => {
+				res = Err(e);
+				return Poll::Ready(None)
 			},
 		}
+	});
+
+	read_stream
+		// NB: Cloning `sender` inside `async` block of `for_each_concurrent` renders the whole
+		// thing not `Send`, so we `zip` the message stream with the stream of `sender` clones here.
+		.zip(stream::repeat(sender))
+		// FIXME: The backlog size here is the same as in PVF host queues. It should either use a
+		// common constant, or another appropriate value should be chosen
+		.for_each_concurrent(30, |message_and_sender| {
```

Comment:
Another thing is that `for_each_concurrent` is not equivalent with calling `ctx.spawn` several times, because 
>the closure are run concurrently (but not in parallel– this combinator does not introduce any threads).

ctx.spawn gives us the opportunity if this being possibly run in parallel, not sure if it really matters, it might not.

Reply:
Speculatively, it has to be okay. What it actually does under the hood is to send a payload to the PVF host and then wait for the answer. All the actual processing is done by the PVF host within its own threading model. But I'm not 100% sure I'm not missing something not obvious here.

Reply:
hmm, it also does code decompression, which can take a couple of ms: https://github.com/paritytech/polkadot-sdk/issues/599.
So that means that one future will block all other from making progress, since they are not being spawned on new tasls

Reply:
Hmm, fair enough. And now we have a dilemma. Go back to the `FuturesUnordered` and add a semaphore, or move the decompression to the validation host (something we want to do anyway) and leave this PR as-is.

Reply:
Your call :D.

Reply:
Benchmarked the decompression with the Moonbeam runtime (a large one), it takes 4.33 msec, not the end of the world, but still significant.

As far as I understand the problem, moving the decompression to the PVF host side (which requires some effort by itself) is only meaningful if we run the PVF host on the blocking pool? And if we ever switch to non-blocking (which we want to do), we'll still have the same problem with some executor threads being blocked in decompression?

If that's true, it's better not to move it but spawn it as a blocking task on the subsystem side. I benchmarked that just in case; it doesn't introduce any overhead and is a future-proof solution.

Reply:
sounds good to me. or we can just use regular spawn like before and see if https://github.com/paritytech/polkadot-sdk/issues/599 is worth taking a look at separately.

maybe it's too much of a micro-optimisation that doesn't make any difference

Reply:
> And if we ever switch to non-blocking (which we want to do), we'll still have the same problem with some executor threads being blocked in decompression?

> If that's true, it's better not to move it but spawn it as a blocking task on the subsystem side.

Sounds good to me too. I do think the PVF host should be non-blocking as it does no blocking work itself. It waits on the child processes but that shouldn't block.